File: perf/zmq-throughput-a.py
import optparse
import sys
import time
import zmq

parser = optparse.OptionParser()
parser.add_option("--address", default="tcp://127.0.0.1:5555", help="Local address.  Default: %default")
parser.add_option("--size", type="int", default=1, help="Message size.  Default: %default")
parser.add_option("--count", type="int", default=100000, help="Message count.  Default: %default")
(options, arguments) = parser.parse_args()

context = zmq.Context()
input = context.socket(zmq.PULL)
input.bind(options.address)

# Don't start timing until we receive our first message ...
message = input.recv()
if len(message) != options.size:
  raise Exception("message size mismatch")

start = time.time()

for i in range(options.count - 1):
  message = input.recv()
  if len(message) != options.size:
    raise Exception("message size mismatch")

elapsed = time.time() - start
throughput = options.count / elapsed
megabits = (throughput * options.size * 8) / 1000000

sys.stdout.write("%s,%s,%s,%s,%s\n" % (elapsed, options.size, options.count, throughput, megabits))

File: perf/zmq-throughput-a.cpp
#include "wallclock.h"

#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <zmq.hpp>

int main(int argc, char* argv[])
{
  if(argc != 4)
    throw std::runtime_error("Usage: <address> <size> <count>");
  const std::string address = argv[1];
  const int size = ::atoi(argv[2]);
  const int count = ::atoi(argv[3]);

  zmq::context_t context(1);
  zmq::socket_t input(context, ZMQ_PULL);
  input.bind(address.c_str());

  // Don't start timing until we receive our first message ...
  zmq::message_t message;
  input.recv(&message, 0);
  if(message.size() != size)
    throw std::runtime_error("message size mismatch");

  const double start = wallclock();

  for(int i = 0; i != count - 1; ++i)
  {
    zmq::message_t message;
    input.recv(&message, 0);
    if(message.size() != size)
      throw std::runtime_error("message size mismatch");
  }

  const double elapsed = wallclock() - start;
  const double throughput = count / elapsed;
  const double megabits = (throughput * size * 8.0) / 1000000.0;
  std::cout << elapsed << "," << size << "," << count << "," << throughput << "," << megabits << "\n";

  return 0;
}
